# Tobig's 14기 2주차 Optimization 과제
### Made by 이지용

# Gradient Descent 구현하기

### 1) "..." 표시되어 있는 빈 칸을 채워주세요  
### 2) 강의내용과 코드에 대해 공부한 내용을 적어서 과제를 채워주세요

In [35]:
import pandas as pd
import numpy as np
import random

In [36]:
data = pd.read_csv('assignment_2.csv')
data.head()

Label  bias  experience  salary
0      1     1         0.7   48000
1      0     1         1.9   48000
2      1     1         2.5   60000
3      0     1         4.2   63000
4      0     1         6.0   76000

## Train Test 데이터 나누기
### 데이터셋을 train/test로 나눠주는 메소드  
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

In [37]:
from sklearn.model_selection import train_test_split

In [38]:
X_train, X_test, y_train, y_test = train_test_split(data.iloc[:, 1:], data.iloc[:, 0], test_size=0.25, random_state = 0)

In [39]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((150, 3), (50, 3), (150,), (50,))

## Scaling  

experience와 salary의 단위, 평균, 분산이 크게 차이나므로 scaler를 사용해 단위를 맞춰줍니다. 

In [40]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
bias_train = X_train["bias"]
bias_train = bias_train.reset_index()["bias"]
X_train = pd.DataFrame(scaler.fit_transform(X_train), columns = X_train.columns)
X_train["bias"] = bias_train
X_train.head()

bias  experience    salary
0     1    0.187893 -1.143335
1     1    1.185555  0.043974
2     1   -0.310938 -0.351795
3     1   -1.629277 -1.341220
4     1   -1.308600  0.043974

이때 scaler는 X_train에 fit 해주시고, fit한 scaler를 X_test에 적용시켜줍니다.  
똑같이 X_test에다 fit하면 안돼요!

In [41]:
bias_test = X_test["bias"]
bias_test = bias_test.reset_index()["bias"]
X_test = pd.DataFrame(scaler.transform(X_test), columns = X_test.columns)
X_test["bias"] = bias_test
X_test.head()

bias  experience    salary
0     1   -1.344231 -0.615642
1     1    0.508570  0.307821
2     1   -0.310938  0.571667
3     1    1.363709  1.956862
4     1   -0.987923 -0.747565

In [42]:
# parameter 개수
N = len(X_train.loc[0])

In [43]:
# 초기 parameter들을 임의로 설정해줍니다.
parameters = np.array([random.random() for i in range(N)])
parameters

array([0.96726719, 0.37331064, 0.11774316])

### * LaTeX   

Jupyter Notebook은 LaTeX 문법으로 수식 입력을 지원하고 있습니다.  
http://triki.net/apps/3466  
https://jjycjnmath.tistory.com/117

## Logistic Function

## $p = 1/(1+e^{−𝑋𝜃})$

In [44]:
def logistic(X, parameters):
    z = 0
    for i in range(len(parameters)) :
        z += parameters[i]*X[i]
    p = 1/(1+np.exp(-z))
    
    return p

In [45]:
logistic(X_train.iloc[1], parameters)

0.8045573467928063

## Object Function

Object Function : 목적함수는 Gradient Descent를 통해 최적화 하고자 하는 함수입니다.  
로지스틱 회귀의 목적함수를 작성해주세요
## $l(p) =-\sum\{y_ilogp+(1-y_i)log(1-p)\}$

In [46]:
def cross_entropy_i(X, y, parameters) :
    p = logistic(X, parameters)              # 위에서 작성한 함수를 활용하세요
    loss = (y*np.log(p)+(1-y)*np.log(1-p))
    return loss

In [47]:
def cross_entropy(X_set, y_set, parameters) :
    loss = 0
    for i in range(X_set.shape[0]):
        X = X_set.iloc[i, :]
        y = y_set.iloc[i]
        loss += cross_entropy_i(X, y, parameters)
    return -loss

In [48]:
cross_entropy(X_test, y_test, parameters)

58.9528669996735

## Gradient of Cross Entropy

## ${\partial\over{\partial \theta_j}}l(p)= -\sum(y_i-p)x_{ij}$

In [49]:
# cross_entropy를 theta_j에 대해 미분한 값을 구하는 함수
def get_gradient_ij_cross_entropy(X, y, parameters, j):
    p = logistic(X, parameters)
    gradient = -((y-p)*X[j].sum())
    return gradient

In [50]:
get_gradient_ij_cross_entropy(X_train.iloc[0, :], y_train.iloc[0], parameters, 1)

-0.05420292075340681

## Batch Gradient Descent  

Batch Gradient Descent : 학습 한 번에 모든 데이터셋에 대해 기울기를 구한다.

In [51]:
def get_gradients_bgd(X_train, y_train, parameters) :
    gradients = [0 for i in range(len(parameters))]
    
    for i in range(len(X_train)):  ## 모든 데이터셋에 대하여 기울기를 구하므로 행의 크기만큼 반복문 필요
        X = X_train.iloc[i, :]
        y = y_train.iloc[i]
        for j in range(len(parameters)):
            gradients[j] += get_gradient_ij_cross_entropy(X, y, parameters, j)
            
    return gradients

In [52]:
gradients_bgd = get_gradients_bgd(X_train, y_train, parameters)
gradients_bgd

[65.26839629122074, -4.707506947891669, 25.42748495236561]

## Stochastic Gradient Descent  

Stochastic Gradient Descent : 학습 한 번에 임의의 데이터에 대해서만 기울기를 구한다.

In [53]:
def get_gradients_sgd(X_train, y, parameters) :
    gradients = [0 for i in range(len(parameters))]
    r = int(random.random()*len(X_train))  ## 임의의 데이터셋 하나의 인덱스만 필요하므로 0~n중에 랜덤하게 뽑기 위해 데이터 길이를 곱해줌
    X = X_train.iloc[r, :]
    y = y_train.iloc[r]
        
    for j in range(len(parameters)):
        gradients[j] = get_gradient_ij_cross_entropy(X, y, parameters, j)
        
    return gradients

In [54]:
gradients_sgd = get_gradients_sgd(X_train, y_train, parameters)
gradients_sgd

[0.5770890888298423, -0.734616758789803, -0.8882005564435079]

## Update Parameters  

In [55]:
def update_parameters(parameters, gradients, learning_rate) :
    for i in range(len(parameters)) :
        gradients[i] *= learning_rate  ## 학습률을 하나씩 곱해야해서 반복문으로 돌리는 것
    parameters -= gradients
    return parameters

In [56]:
update_parameters(parameters, gradients_bgd, 0.01)

array([ 0.31458322,  0.42038571, -0.13653169])

## Gradient Descent  

위에서 작성한 함수들을 조합해서 Gradient Descent를 진행하는 함수를 완성해주세요

learning_rate = 학습률  
max_iter = 최대 반복 횟수  
tolerance = step 이동이 무의미할 시 중단 조건

In [57]:
def gradient_descent(X_train, y_train, learning_rate=0.01, max_iter=100000, tolerance=0.0001, optimizer="bgd") :
    count = 1
    point = 100 if optimizer == "bgd" else 10000  ## point에 도달할 때마다 학습을 중단해야 할지 점검하기 위한 지점
    N = len(X_train.iloc[0])
    parameters = np.array([random.random() for i in range(N)])
    gradients = [0 for i in range(N)]
    loss = 0
    
    while count < max_iter :
        
        if optimizer == "bgd" :
            gradients = get_gradients_bgd(X_train, y_train, parameters)
        elif optimizer == "sgd" :
            gradients = get_gradients_sgd(X_train, y_train, parameters)
            # loss, 중단 확인
        if count%point == 0 :
            new_loss = cross_entropy(X_train, y_train, parameters)
            print(count, "loss: ",new_loss, "params: ", parameters, "gradients: ", gradients)
            
            #중단 조건
            if abs(new_loss-loss) < tolerance/len(y_train) : ## step의 이동이 더이상 의미없을 때, 즉 loss 값의 변동이 극히 작을 때 중단
                break
            loss = new_loss
                
            
                
        parameters = update_parameters(parameters, gradients, learning_rate)
        count += 1
    return parameters

In [58]:
new_param_bgd = gradient_descent(X_train, y_train)
new_param_bgd

100 loss:  45.34382296890982 params:  [-1.63080242  3.49713612 -3.32195088] gradients:  [0.26749435207169003, -0.8796193985760208, 0.821816857490237]
200 loss:  44.796957196080434 params:  [-1.78402755  3.99778559 -3.78729337] gradients:  [0.08117329288182086, -0.26374123929981536, 0.2438797218408273]
300 loss:  44.73979724502206 params:  [-1.83482536  4.16258409 -3.9394377 ] gradients:  [0.029286796375830218, -0.09487102111523638, 0.08744442911973072]
400 loss:  44.73206145253853 params:  [-1.85364447  4.22351504 -3.99556646] gradients:  [0.011140461925200393, -0.03605096342926666, 0.03319010447485067]
500 loss:  44.730926353032494 params:  [-1.86087105  4.24689604 -4.01708729] gradients:  [0.004319077450530895, -0.01397130479509539, 0.012856884805586229]
600 loss:  44.730754834614245 params:  [-1.86368277  4.25599071 -4.02545581] gradients:  [0.0016866020873587684, -0.005454990242893851, 0.005019008568629663]
700 loss:  44.73072862593418 params:  [-1.86478227  4.25954673 -4.02872752]

array([-1.86538198,  4.26148621, -4.03051184])

## Hyper Parameter Tuning

Hyper Parameter들을 매번 다르게 해서 학습을 진행해 보세요. 다른 점들을 발견할 수 있습니다.

In [59]:
new_param_sgd = gradient_descent(X_train, y_train, learning_rate=0.01, max_iter=100000, tolerance=0.0001, optimizer="sgd")
new_param_sgd

10000 loss:  46.566150611712494 params:  [-1.48015943  2.98794247 -2.97642657] gradients:  [0.0315230947433199, -0.023280043145461705, -0.0027724197017444563]
20000 loss:  45.22598610387789 params:  [-1.80570804  3.79609453 -3.46108401] gradients:  [-0.7089966080074461, -0.966865170457979, -0.8729760363699444]
30000 loss:  44.824287123695065 params:  [-1.72269429  4.07366769 -3.83298193] gradients:  [0.3165290704908505, 0.5782699509424195, 0.5358885346829997]
40000 loss:  44.86891536798804 params:  [-1.69729517  4.13342361 -3.91259882] gradients:  [0.16249207208203628, -0.1431605049925162, -0.15362813918245]
50000 loss:  44.744931549010886 params:  [-1.87928988  4.20204878 -3.95865675] gradients:  [-0.6804647850906472, 0.5510193721978272, 0.7779991747183738]
60000 loss:  44.736239580450764 params:  [-1.87718671  4.32723472 -4.07068442] gradients:  [0.0014067523685090432, -0.0016403816358342724, -0.00012372224280919074]
70000 loss:  44.87301985087269 params:  [-1.82329766  4.26619929 -4

array([-1.89892211,  4.19774877, -3.99997805])

## Predict Label

In [60]:
y_predict = []
for i in range(len(y_test)):
    p = logistic(X_test.iloc[i,:], new_param_bgd)
    if p> 0.5 :
        y_predict.append(1)
    else :
        y_predict.append(0)

## Confusion Matrix

In [61]:
from sklearn.metrics import *
tn, fp, fn, tp = confusion_matrix(y_test, y_predict).ravel()
confusion_matrix(y_test, y_predict)

array([[38,  2],
       [ 1,  9]], dtype=int64)

In [62]:
accuracy_score(y_test, y_predict)

0.94

In [63]:
f1_score(y_test, y_predict)

0.8571428571428572

정확도는 0.94, f1 score는 0.857로, TN의 비중이 높아 imbalance로 인해 f1 score가 조금 낮게 나오긴 했으나, 전반적으로 잘 예측했다고 평가할 수 있음.